src/H5FDint.c

```diff
@@ -260,152 +290,2049 @@ H5FD_write(H5FD_t *file, H5FD_mem_t type, haddr_t addr, size_t size, const void
+
+    /* Adjust for base address in file (convert to relative address)  */
+    ret_value -= file->base_addr;
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5FD_get_eof() */
+
+/*-------------------------------------------------------------------------
+ * Function:     H5FD_driver_query
+ *
+ * Purpose:      Similar to H5FD_query(), but intended for cases when we don't
+ *               have a file available (e.g. before one is opened). Since we
+ *               can't use the file to get the driver, the driver is passed in
+ *               as a parameter.
+ *
+ * Return:       SUCCEED/FAIL
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5FD_driver_query(const H5FD_class_t *driver, unsigned long *flags /*out*/)
+{
+    herr_t ret_value = SUCCEED; /* Return value */
+
+    FUNC_ENTER_NOAPI_NOINIT_NOERR
+
+    HDassert(driver);
+    HDassert(flags);
+
+    /* Check for the driver to query and then query it */
+    if (driver->query)
+        ret_value = (driver->query)(NULL, flags);
+    else
+        *flags = 0;
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5FD_driver_query() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5FD_sort_vector_io_req
+ *
+ * Purpose:     Determine whether the supplied vector I/O request is
+ *              sorted.
+ *
+ *              if is is, set *vector_was_sorted to TRUE, set:
+ *
+ *                 *s_types_ptr = types
+ *                 *s_addrs_ptr = addrs
+ *                 *s_sizes_ptr = sizes
+ *                 *s_bufs_ptr = bufs
+ *
+ *              and return.
+ *
+ *              If it is not sorted, duplicate the type, addrs, sizes,
+ *              and bufs vectors, storing the base addresses of the new
+ *              vectors in *s_types_ptr, *s_addrs_ptr, *s_sizes_ptr, and
+ *              *s_bufs_ptr respectively.  Determine the sorted order
+ *              of the vector I/O request, and load it into the new
+ *              vectors in sorted order.
+ *
+ *              Note that in this case, it is the callers responsibility
+ *              to free the sorted vectors.
+ *
+ *                                            JRM -- 3/15/21
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ *-------------------------------------------------------------------------
+ */
+
+static int
+H5FD__vsrt_tmp_cmp(const void *element_1, const void *element_2)
+{
+    haddr_t addr_1    = ((const H5FD_vsrt_tmp_t *)element_1)->addr;
+    haddr_t addr_2    = ((const H5FD_vsrt_tmp_t *)element_2)->addr;
+    int     ret_value = 0; /* Return value */
+
+    FUNC_ENTER_STATIC_NOERR
+
+    /* Sanity checks */
+    HDassert(H5F_addr_defined(addr_1));
+    HDassert(H5F_addr_defined(addr_2));
+
+    if (H5F_addr_gt(addr_1, addr_2)) {
+
+        ret_value = 1;
+    }
+    else if (H5F_addr_lt(addr_1, addr_2)) {
+
+        ret_value = -1;
+    }
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* H5FD__vsrt_tmp_cmp() */
```

Comment:
Stop the madness. Write simple code simply:

```C
static int
vsrt_tmp_cmp(const void *element_1, const void *element_2)
{
    haddr_t addr_1    = ((const H5FD_vsrt_tmp_t *)element_1)->addr;
    haddr_t addr_2    = ((const H5FD_vsrt_tmp_t *)element_2)->addr;

    if (addr_1 < addr_2)
        return -1;
    else if (addr_1 == addr_2)
        return 0;
    else
        return 1;
}
```

Reply:
I won't argue about the extra whitespace (this is John's code), but I don't think throwing away the ENTER/LAEVE macros is a good idea. If this one is adding overhead in production mode with all tracing, etc disabled we should figure out why that is instead of just not using the macros that are supposed to be used everywhere.

Reply:
Also we definitely shouldn't throw away the asserts

Reply:
The H5F_addr_*() macros do other range-checking besides the comparison operation.  Please keep them in.

Reply:
Looking at FUNC_ENTER_STATIC_NOERR it should resolve to nothing if NDEBUG is defined and the codestack is disabled.

Reply:
If we expand the macros in the original, then we get this:

```C
static int
H5FD__vsrt_tmp_cmp(const void *element_1, const void *element_2)
{   
    haddr_t addr_1    = ((const H5FD_vsrt_tmp_t *)element_1)->addr;
    haddr_t addr_2    = ((const H5FD_vsrt_tmp_t *)element_2)->addr;
    int     ret_value = 0; /* Return value */

    FUNC_ENTER_STATIC_NOERR

    /* Sanity checks */
    HDassert(addr_1 != HADDR_UNDEF);
    HDassert(addr_2 != HADDR_UNDEF);

    if (addr_1 != HADDR_UNDEF && addr_2 != HADDR_UNDEF &&
        addr_1 > addr_2) {

        ret_value = 1;
    }
    else if (addr_1 != HADDR_UNDEF && addr_2 != HADDR_UNDEF &&
        addr_1 < addr_2) {

        ret_value = -1;
    }

    FUNC_LEAVE_NOAPI(ret_value)
} /* H5FD__vsrt_tmp_cmp() */
```

If NDEBUG is *not* \#defined, then in effect the code is like this:

```C
static int
H5FD__vsrt_tmp_cmp(const void *element_1, const void *element_2)
{   
    haddr_t addr_1    = ((const H5FD_vsrt_tmp_t *)element_1)->addr;
    haddr_t addr_2    = ((const H5FD_vsrt_tmp_t *)element_2)->addr;
    int     ret_value = 0; /* Return value */

    FUNC_ENTER_STATIC_NOERR

    /* Sanity checks */
    if (addr_1 == HADDR_UNDEF) { /* print reason */ abort(); }
    if (addr_2 == HADDR_UNDEF) { /* print reason */ abort(); }

    if (addr_1 > addr_2) {

        ret_value = 1;
    }
    else if (addr_1 < addr_2) {

        ret_value = -1;
    }

    FUNC_LEAVE_NOAPI(ret_value)
} /* H5FD__vsrt_tmp_cmp() */
```

However, if NDEBUG *is* \#defined, then the code is effectively this:

```C
static int
H5FD__vsrt_tmp_cmp(const void *element_1, const void *element_2)
{   
    haddr_t addr_1    = ((const H5FD_vsrt_tmp_t *)element_1)->addr;
    haddr_t addr_2    = ((const H5FD_vsrt_tmp_t *)element_2)->addr;
    int     ret_value = 0; /* Return value */

    FUNC_ENTER_STATIC_NOERR

    /* Sanity checks */
    /* first assert: no effect */
    /* second assert: no effect */

    if (addr_1 != HADDR_UNDEF && addr_2 != HADDR_UNDEF &&
        addr_1 > addr_2) {

        ret_value = 1;
    }
    else if (addr_1 != HADDR_UNDEF && addr_2 != HADDR_UNDEF &&
        addr_1 < addr_2) {

        ret_value = -1;
    }

    FUNC_LEAVE_NOAPI(ret_value)
} /* H5FD__vsrt_tmp_cmp() */
```

The NDEBUG version is not a comparison function. Using that function, `qsort` may not put your array into sorted order if there are any HADDR_UNDEF in it.

Using the simple comparison function, `qsort` always harmlessly moves the HADDR_UNDEF to the end.

Of all of the places to look for HADDR_UNDEF, what makes the comparison function a good one? You can check before the sort or after the sort. Is it important for some reason to check O(count log count) times *during* the sort? BTW, I am *not* particularly bothered by performance impacts this may have. I am bothered by what I see as unquestioned double-, triple-, and quadruple-checking that obscures really simple algorithmic ideas.

Reply:
> The NDEBUG version is not a comparison function. Using that function, qsort may not put your array into sorted order if there are any HADDR_UNDEF in it.

While I have no comments of any substance on the rest of the discussion here, this quirk is something I noticed during my parallel compression updates work and had to work around it. We have H5F_addr_cmp, but that's just a composition of these other macros and exhibits the same issue.

Reply:
Part of the idea for the asserts in this function is to maximize the readability of the function by minimizing the amount the developer reading the function needs to know about the rest of the library. By documenting these assumptions the developer doesn't need to dig around the rest of the code to try to figure them out. By making it as assert instead of a comment we're immediately given a useful error message if they're ever violated.

As to the comment on qsort issues if there are HADDR_UNDEFs, there wont be, and the assert is making sure of that.